native_ops/fbw_op.cu

```diff
@@ -0,0 +1,613 @@
+    for (unsigned seq = 0u; seq < n_seqs; seq++) {
+        std::stringstream filename;
+        filename << "alignment.dump." << batch_idx << '.' << seq;
+        std::ofstream out(filename.str().c_str(),
+                          std::ios::out | std::ios::trunc);
+        for (unsigned t = 0u; t < n_frames; t++) {
+            if (t > 0u && t >= seq_lens[seq]) {
+                break;
+            }
+            float sum = std::numeric_limits<float>::infinity();
+            for (unsigned s = start_states[seq]; s <= end_states[seq]; s++) {
+                const float val = state_buffer[t * n_states + s];
+                float diff = val - sum;
+                if (!isnan(diff)) {
+                    sum = -log1p(exp(-abs(diff))) + fminf(sum, val);
+                }
+            }
+            for (unsigned s = start_states[seq]; s <= end_states[seq]; s++) {
+                const float val = state_buffer[t * n_states + s] - sum;
+                if (val <= pruning) {
+                    out << t << ' ' << (s - start_states[seq]) << ' ' << val
+                        << '\n';
+                }
+            }
+        }
+    }
+}
+
+void write_output_to_file(float *d_out, unsigned *d_seq_lens, float pruning,
+                          unsigned n_frames, unsigned n_seqs,
+                          unsigned n_emissions, unsigned batch_idx) {
+    std::vector<float> buffer(n_frames * n_seqs * n_emissions);
+    std::vector<unsigned> seq_lens(n_seqs);
+
+    HANDLE_ERROR(cudaMemcpy(buffer.data(), d_out, buffer.size() * sizeof(float),
+                            cudaMemcpyDeviceToHost));
+    HANDLE_ERROR(cudaMemcpy(seq_lens.data(), d_seq_lens,
+                            seq_lens.size() * sizeof(unsigned),
+                            cudaMemcpyDeviceToHost));
+
+    for (unsigned seq = 0u; seq < n_seqs; seq++) {
+        std::stringstream filename;
+        filename << "target.dump." << batch_idx << '.' << seq;
+        std::ofstream out(filename.str().c_str(),
+                          std::ios::out | std::ios::trunc);
+        for (unsigned t = 0u; t < n_frames; t++) {
+            if (t > 0u && t >= seq_lens[seq]) {
+                break;
+            }
+            for (unsigned e = 0u; e < n_emissions; e++) {
+                const float val =
+                    buffer[t * n_seqs * n_emissions + seq * n_emissions + e];
+                if (val <= pruning) {
+                    out << t << ' ' << e << ' ' << val << '\n';
+                }
+            }
+        }
+    }
+}
+
+std::vector<torch::Tensor> fbw_cuda(std::vector<torch::Tensor> torch_inputs) {
+    // am_scores, edges, weights, start_end_states, index, state_buffer
+    assert(torch_inputs.size() == 6);
+    Ndarray *am_scores = &torch_inputs[0];
+    Ndarray *edges = &torch_inputs[1];
+    Ndarray *weights = &torch_inputs[2];
+    Ndarray *start_end_states = &torch_inputs[3];
+    Ndarray *seq_lens = &torch_inputs[4];
+    Ndarray *state_buffer = &torch_inputs[5];
```

Comment:
Please allocate the state_buffer inside the kernel.

Reply:
Yes no problem. Only note that we will need an extra `num_states` parameter in order to allocate the buffer.